Reimplement !dumpheap and !verifyheap in C#

leculver opened the pull request:
Changes:

- Add "WriteDmlExec" to IConsoleService.  While we shouldn't add *all* DML based commands to IConsoleService, exec is special in that we can print the right value out when DML is not supported.  This lets IFileLogger and similar non-DML console services print just the text and ignore the underlying command.
- Added [DebugCommand] as a subclass of CommandAttribute.  This allows us to define commands which are only available in debug builds of SOS...which is important for the heap corruption test command I implemented.
- Added `!sos SimulateGCHeapCorruption` (debug only!) to write some values to the GC heap to be able to better test !dumpheap and !verifyheap when heap corruption is present.
- Added a managed implementation of !dumpheap and !verifyheap.
- Added additional DML capabilities to TableOutput.

DumpHeap changes:

- Implemented !dumpheap based on ClrMD/C# code, deleted the old implementation.
- The ExtensionCommand command line parser is not handling a single - properly.  This means `-stat` became `--stat`, because it's the only way it works.  I will either update this PR with the fix (if it's still open) or open a new one as soon as I have a way to fix this.  This is my top priority to fix.
- The `!dumpheap -thinlock` command is not yet implemented!  I added it to my roadmap here: https://github.com/dotnet/diagnostics/issues/3736, and will be working on it as soon as the `--` issue (above) is resolved.

Quick notes:

- I'm waiting on #3721 to complete before merging.
- I still need to fix tests, I wanted to get a PR as I work on tests.

leculver commented:
@mikem8361 There's two todos left here (aside from tests, which will be fixed before merging):

1. Fix the command argument issue.  Right now, we have to use `--stat` instead of `-stat`, for example.  I'm going to fix this as my top priority.
2.  Add -thinlock support.  I want to add the underlying feature to ClrMD instead of having the logic in SOS, so it will take a while to do that and round-trip the feature into this repo.  I'd like to do that as another checkin, if that's ok.  (It's on my roadmap, planning to tackle it as soon as the command argument issue is fixed).

Once PR feedback is fixed, and tests are fixed, am I ok to merge this PR?  Or wait for 1 and 2 above to be fixed before merging?

leculver commented:
I almost forgot, I did change !verifyheap's output and function.  The new code is **very** well tested.  I took a standard process and ran the original verifyheap over it while corrupting every pointer aligned address with the program at the end of this message.

I then took the new ClrHeap.VerifyHeap method and made sure that it reported the same failures for each corruption generated.  The only remaining work item for the new !verifyheap is that we don't validate the GC card table.  I actually found that the old C++ verifyheap _also_ doesn't validate the card table...so there's no behavior change here.  I have card table validation added to my roadmap (#3736).  I'm waiting on Andrew Au to give me some details on GC heap structure to properly read/validate the card table first.

``` C#
    foreach (ClrSegment segment in runtime.Heap.Segments)
    {
        Console.Write($"Segment {segment.Address:x} Length: {segment.ObjectRange.Length:n0} bytes");
        var segWatch = Stopwatch.StartNew();
        for (ulong address = segment.ObjectRange.Start; address < segment.ObjectRange.End; address += (uint)dt.DataReader.PointerSize)
        {
            Add(WriteAndVerify(address, (byte)0xcc));
            Add(WriteAndVerify(address, (byte)0));

            Add(WriteAndVerify(address, (ushort)0xcccc));
            Add(WriteAndVerify(address, (ushort)0));

            Add(WriteAndVerify(address, (uint)0xcccccccc));
            Add(WriteAndVerify(address, (uint)0));

            if (dt.DataReader.PointerSize == 8)
            {
                Add(WriteAndVerify(address + 4, (byte)0xcc));
                Add(WriteAndVerify(address + 4, (byte)0));

                Add(WriteAndVerify(address + 4, (ushort)0xcccc));
                Add(WriteAndVerify(address + 4, (ushort)0));

                Add(WriteAndVerify(address + 4, (uint)0xcccccccc));
                Add(WriteAndVerify(address + 4, (uint)0));
            }

            if (sw.Elapsed.Seconds >= 30)
            {
                File.WriteAllText(output, JsonSerializer.Serialize(failures, new JsonSerializerOptions() { WriteIndented = true }));
                sw.Restart();
            }
        }

        Console.WriteLine($" completed in {segWatch.Elapsed}");
    }
```

mikem8361 commented:
I'm ok with not doing both of those.  I thought I the single dash aliases in the `sosstatus` and `runtimes` commands (maybe more) worked but I haven't tried them in a while.  The parser (System.CommandLine) is probably doing some special for single dash commands like assuming they are single letter aliases that can be combined like Unix command line options. 

But the biggest thing you need to wait for is Juan's PR and you will have a lot work merging and fixing all the analysis errors in your new code (we are enforcing a coding convention that includes things like requiring { } around single-line conditional bodies, etc.).

leculver commented on src/Microsoft.Diagnostics.ExtensionCommands/DumpHeapCommand.cs:
```diff
@@ -0,0 +1,291 @@
+            }
+        }
+
+        private string Sanitize(string str, int maxLen)
+        {
+            IEnumerable<char> chars = str.Take(maxLen);
+
+            if (chars.All(char.IsLetterOrDigit))
+                return str;
+
+            return new(chars.Select(c => (char.IsLetterOrDigit(c) || char.IsPunctuation(c) || c == ' ') ? c : StringReplacementCharacter).ToArray());
+        }
```
Good catch!  I'll use this.  :)

leculver commented on src/Microsoft.Diagnostics.DebugServices.Implementation/FileLoggingConsoleService.cs:
```diff
@@ -158,6 +158,22 @@ public void WriteDml(string text)
+                {
+                    writer.Write(text);
+                }
+                catch (Exception ex) when (ex is IOException || ex is ObjectDisposedException || ex is NotSupportedException)
```
I was matching the other methods in the file, but now that they've been updated by Juan's PR, this is fixable now.

leculver commented:
I rebased this on top of @hoyosjs's diagnostic code cleanup with the exception of `IDE0011`.  IDE0011 is brutal, the extra brace requirements _severely_ reduces screen real-estate, sometimes by more than 1/3rd of what you can see previously.  I have it temporarily disabled in this PR, as I'm hoping you all will reconsider that one.  =)  I sent email to Juan/Mike, we don't have to hash it out in this PR.

I've built thinlock support into ClrMD, I'll have that and test fixes ready by monday.

leculver commented:
Well, this was actually just a bug in my code:

> Fix the command argument issue. Right now, we have to use --stat instead of -stat, for example. I'm going to fix this as my top priority.

I had put `-s` as an alias for `--segment`, which System.Console.CommandLine gets confused when you have a `-stat` command.  The `-stat` gets interpreted as `-s tat`, which obviously is not what was intended.

In any case, I've added -thinlock and fixed tests.  This is ready for code review.  Obviously we have to figure out the right solution to `IDE0011` before checkin, which we'll discuss offline.

mikem8361 commented on src/Microsoft.Diagnostics.ExtensionCommands/LiveObjectService.cs:
```diff
@@ -0,0 +1,93 @@
+    {
+        public int UpdateSeconds { get; set; } = 15;
+
+        private ObjectSet _liveObjs;
```
Nit: I usually put private fields first in a class then protected and then public fields/properties.

mikem8361 commented on src/SOS/SOS.Extensions/ConsoleServiceFromDebuggerServices.cs:
```diff
@@ -27,12 +28,20 @@ public ConsoleServiceFromDebuggerServices(DebuggerServices debuggerServices)
 
         public void WriteDml(string text) => _debuggerServices.OutputDmlString(DEBUG_OUTPUT.NORMAL, text);
 
+        public void WriteDmlExec(string text, string cmd)
```
Does the `dumpasync` command need to use this?  When Stephen Toub rewrote it, he added the DML support.  You don't have to do in your PR if this is necessary, but we should track it.

leculver replied:
Yep, I plan to update !dumpasync when I figure submit changes to fix DML output not showing up for extension commands.  I'll tackle that with the DML checkin.

mikem8361 commented on src/SOS/SOS.UnitTests/Scripts/GCPOH.script:
```diff
@@ -39,7 +39,7 @@ VERIFY:.*Thread <HEXVAL>:
 
 SOSCOMMAND:VerifyHeap
-VERIFY:\s*No heap corruption detected.\s*
+VERIFY:.* 0 errors.*
```
Isn't "No heap corruption detected" a better message for most users?

leculver replied:
Good idea.  I will add that back as an extra line if we find no errors.

leculver commented:
@mikem8361 This is ready for checkin.  All tests are passing, DML is working, arguments are fixed, and code style is matching the rest of diagnostics.

Any other feedback or reviews needed?

mikem8361 commented:
I still have a lot to review.  I started with the simple things.  The main command files are still left.  I'm off for the day but I may be able to do a little this morning.

leculver commented:
It can wait if you are out of office today.  I'll start on the next bit of work with this as the base branch.  It won't be a problem to rebase after this is merged.

mikem8361 commented on src/Microsoft.Diagnostics.ExtensionCommands/DumpHeapCommand.cs:
```diff
@@ -0,0 +1,381 @@
+            FilteredHeap.SortSegments = (seg) => seg.OrderBy(seg => seg.Start);
+        }
+
+        private static bool ParseHexString(string str, out ulong value)
```
The CommandBase has a TryParseAddress function that does the same thing.  If the CommandBase has limitations or problems that your implementation fixes could you move them to CommandBase.TryParseAddress and use that one.  

Someday I plan to add a expression evaluation service (it will call the underlying host debugger's expression evaluator) that will hook into the command service parser that makes all of this moot. Using this CommandBase function will help me find all the places the manual parsing is done.

leculver replied:
Thanks, I missed that.  I'll make that change.

It would be really nice if that was part of the CommandAttribute.  I want to write this:

``` C#
[Option(Name="Foo", Format="HexAddress")]
public ulong Address { get; set }
```

But as I was poking around System.CommandLine it wasn't obvious how set a custom parser like that, so I've left it for another day.

mikem8361 replied:
There is a way to hook SystemCommandLine option/argument parsing theoretically. I haven't tried it yet but it is on my list as part of an official IExpressionService.  And yes, adding hint properties to the option attr is a good idea.

leculver replied:
I'm going to tackle this on my next PR.  I want to get this merged so I can rebase my other PR without too much drift.

mikem8361 reviewed (approved):
Thanks for doing of this.  This is great!
